dipy/reconst/ivim.py

```diff
@@ -0,0 +1,380 @@
+        # if there is noise in the data the estimated x0 might not be feasible.
+        # In such a case we will use the values for the parameters
+        # from the bounds set with `bounds_check`.
+        if self.bounds is None:
+            bounds_check = [(0., 0., 0., 0.), (np.inf, 1., 0.1, 0.1)]
+        else:
+            bounds_check = self.bounds
+
+        x0 = np.where(x0 > bounds_check[0], x0, bounds_check[0])
+        x0 = np.where(x0 < bounds_check[1], x0, bounds_check[1])
+
+        return x0
+
+    @multi_voxel_fit
+    def fit(self, data, mask=None):
+        """ Fit method of the Ivim model class
+
+        Parameters
+        ----------
+        data : array
+            The measured signal from one voxel. A multi voxel decorator
+            will be applied to this fit method to scale it and apply it
+            to multiple voxels.
+
+        mask : array
+            A boolean array used to mark the coordinates in the data that
+            should be analyzed that has the shape data.shape[:-1]
+
+        Returns
+        -------
+        IvimFit object
+
+        References
+        ----------
+        .. [1] Federau, Christian, et al. "Quantitative measurement
+                   of brain perfusion with intravoxel incoherent motion
+                   MR imaging." Radiology 265.3 (2012): 874-881.
+        """
+        # Call the function estimate_S0_D to get initial x0 guess.
+        x0 = self.estimate_x0(data)
+        # Use leastsq to get ivim_params
+        params_in_mask = self._leastsq(data, x0)
+        return IvimFit(self, params_in_mask)
+
+    def predict(self, ivim_params, gtab, S0=1.):
+        """
+        Predict a signal for this IvimModel class instance given parameters.
+
+        Parameters
+        ----------
+        ivim_params : array
+            The ivim parameters as an array [S0, f, D_star and D]
+
+        Returns
+        -------
+        ivim_signal : array
+            The predicted IVIM signal using given parameters.
+        """
+        return ivim_prediction(ivim_params, gtab)
+
+    def _estimate_S0_D(self, data):
+        """
+        Obtain initial guess for S0 and D for two stage fitting.
+
+        Using TensorModel from reconst.dti, we fit an exponential
+        for signal values which are greater than a particular bvalue
+        (split_b). The apparent diffusion coefficient gives us an
+        initial estimate for D and S0.
+
+        Parameters
+        ----------
+        data : array
+            The measured signal.
+
+        Returns
+        -------
+        S0_hat : float
+            Initial S0 guess.
+
+        D_guess : float
+            Initial D_guess.
+        """
+        gtab = self.gtab
+        split_b = self.split_b
+        bvals_ge_split = gtab.bvals[gtab.bvals > split_b]
+        bvecs_ge_split = gtab.bvecs[gtab.bvals > split_b]
+        gtab_ge_split = gradient_table(bvals_ge_split, bvecs_ge_split.T)
+        tensor_model = TensorModel(gtab_ge_split)
```

Comment:
We haven't written this to deal with tensor data yet. This would be a good comment to add in the documentation though explaining that the user needs to average the directions or select specific ones.